### PR TITLE
Add Pex support for ray-on-spark, this allows users to 'ship' any ver…

### PR DESCRIPTION
**Pull request by GabeChurch**
…sion of ray to any spark cluster using the only universally compatible mechanism, pex

## Why are these changes needed?
PEX is a very powerful python dependency distribution tool. It especially shines when it is used in conjunction with PySpark. It is currently the only dependency distribution utility capable of being used with any spark cluster, allowing users to "bootstrap" python dependencies isolated "per-SparkSession". In it's existing form, ray-on-spark is not capable of being deployed and executed from a .pex file on a spark cluster. This PR changes that and makes ray-on-spark fully portable with .pex. This has numerous benefits (multiple ray versions supported on a single spark cluster for complex dependency chains, lighter-weight more generic spark images, easier per-job python dependency resolution, and more).

These changes primarily impact the way subprocesses are called within Ray node and service launch scripts, when a specific flag is set.  

From the user-facing side the only change is a default=False option under ray.util.spark.setup_ray_cluster. If users add "uses_pex=True" the cluster-init will first check for an environment variable called "RAY_PEX_FILE" which must be set on the spark driver (where the cluster_init is running). This should be a local path to the pex file like "~./ray-pex-deps.pex". It can be distributed to the spark workers using spark.files (s3_location or local_location), from that point a second check for that environment variable runs for the spark workers. It must be configured in the spark.conf like spark.executorEnv.RAY_PEX_FILE. If it is set, it will be used as the python executable instead of sys.executable. 

#### Generate Ray Pex File using future version of ray
`/opt/conda/bin/pex ray[default,rllib,tune]==2.10.1 -o $(pwd)/ray-pex-deps.pex --inherit-path=fallback`

#### Initialize PySpark SparkSession
(Lots of options here, could use spark-submit, google-sparksubmit-operator for kubernetes, python kernel in Jupyter, etc)
 
For Jupyter you would do below individually in notebook cells -- note require [pants-jupyter-plugin](https://pypi.org/project/pants-jupyter-plugin/#description)
`%reload_ext pants_jupyter_plugin`
`%pex_load ray-pex-deps.pex`
 
In a python script or notebook cell
 ```
import pyspark
from pyspark import SparkContext, SparkConf
from pyspark.sql import SparkSession
import os

pyspark_deps_name = 'ray-pex-deps.pex'
os.environ['PYSPARK_PYTHON'] = f"./{pyspark_deps_name}"
#required for ray on spark to utilize pex file for subprocesses
os.environ['RAY_PEX_FILE'] = f"./{pyspark_deps_name}"

conf = SparkConf().setMaster("spark://spark-master:7077") # or for local .setMaster('local[*]')
hostname = os.popen("hostname -i").read().split("\n")[0]
conf.set("spark.deployMode","client")
conf.set("spark.driver.host",hostname)
#shipping pex dependencies from local or s3, the pex file must exist locally for most use cases
conf.set("spark.files",pyspark_deps_name)
#required for ray on spark to utilize pex file for subprocesses
conf.set("spark.executorEnv.RAY_PEX_FILE", f"./{pyspark_deps_name}")

sc = pyspark.SparkContext(conf=conf)
sc.setLogLevel("warn")
spark = SparkSession(sc)

from ray.util.spark import setup_ray_cluster, shutdown_ray_cluster
conn_str = setup_ray_cluster(uses_pex=True)

import ray
ray.init(conn_str[0])

#Ray code below
#End ray code
shutdown_ray_cluster()
spark.stop()
 ```
 You could launch via a python script on the spark cluster like `ray-pex-deps.pex my_example_python.py`

I'm planning on posting some material and examples once this is merged, happy to share some links. 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [x] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
          `No doc identified`
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by GabeChurch** [comment]
CC @jjyao if can review

**Comment by jjyao** [comment]
@WeichenXu123 could you review it as well?

**Comment by WeichenXu123** [comment]
Sure!

File: python/ray/util/spark/cluster_init.py [hunk 1]
```diff
@@ -1851,10 +1889,21 @@ def start(
...
+                f"{ray_path[0]}/util/spark/start_ray_node.py",
+            ]
+        else:
+            ray_head_node_cmd = [
+                sys.executable,
+                "-m",
+                "ray.util.spark.start_ray_node",
+            ]
+
+        ray_head_node_cmd.extend([
```
Comment (WeichenXu123) [inline python/ray/util/spark/cluster_init.py:1906]: Another case: autoscaling is disabled. You need to update that case Ray head node command.

Comment (GabeChurch) [reply python/ray/util/spark/cluster_init.py:1906]: Ah I see your point, I will modify+test+push to handle that

Comment (GabeChurch) [reply python/ray/util/spark/cluster_init.py:1910]: @WeichenXu123 please take a look again, I made sure to trigger both autoscaling and fixed number of workers after making the changes. 

image

Comment (WeichenXu123) [reply python/ray/util/spark/cluster_init.py:1910]: I think this place you need to update it too:

https://github.com/ray-project/ray/blob/6e10ca79bb623a0004c8ad5300a3f5d8aa2c4eb3/python/ray/util/spark/cluster_init.py#L682

Does your test cover it ?

**Review by WeichenXu123 — COMMENTED** [review]
Overall good for Ray-on-Spark part code, except one comment.

@jjyao Could you review / approve other parts code ?

**Review by WeichenXu123 — COMMENTED** [review]
Rethink:

I am considering in Ray-on-Spark `start_ray_node.py`, avoid using Ray CLI command, instead, directly invoke `ray.scripts.scripts.main()`

The reason is ,

in Databricks runtime, we face one issue that Ray CLI script Shebang is hardcoded to a fixed python binary program,
but our Databricks allow user to create python environment layer on top of base python environment. Then Ray tasks can't use python packages installed in user cusomized python environment layer.

So would you mind removing Ray-on-Spark part changes ?

Then I will file a PR to change `start_ray_node.py` to invoke `ray.scripts.scripts.main()` (and this will make it supporting PEX too)  , or if @GabeChurch you are interested, you can file PR for it, and ping me to review.

CC @jjyao

File: python/ray/_private/services.py [hunk 2]
```diff
@@ -2179,17 +2245,29 @@ def start_ray_client_server(
...
-        setup_worker_path,
-        "-m",
-        "ray.util.client.server",
+
+    if RAY_USES_PEX:
+        import ray
+        ray_path = ray.__path__
+        command = [
+            RAY_PEX_FILE,
+            f"{ray_path[0]}/util/client/server/server.py",
```
Comment (jjyao) [inline python/ray/_private/services.py:2254]: I'm not familiar with pex but why do we do this differently here than other places where only sys.executable is replaced?

**Comment by anyscalesam** [comment]
@GabeChurch  have you had a chance to review the comments from @WeichenXu123 and @jjyao ?

**Comment by WeichenXu123** [comment]
Update:

> in Databricks runtime, we face one issue that Ray CLI script Shebang is hardcoded to a fixed python binary program,

This issue has been fixed in https://github.com/ray-project/ray/pull/44827

**Comment by GabeChurch** [comment]
> @GabeChurch  have you had a chance to review the comments from @WeichenXu123 and @jjyao ?

Yes @anyscalesam and @jjyao I took the feedback and made some changes, I will tag you and @WeichenXu123 in a new P/R later today that I am finalizing. With those changes I have successfully added inherent support for pex (no extra config required) and incorporated the suggestion of invoking ray scripts.scripts directly as initially mentioned by @WeichenXu123. The new changes have an added benefit of reducing the overhead of a nested subproceses with a direct fork (subsequently eliminating an extra monitoring thread process), add support for configurable logging levels in ray-on-spark with more robust debug logging info, and, as mentioned allow ray-on-spark to be run from any kind of virtualenv (including pex). I will update this comment with the linked changes and close this P/R subsequently.

**Comment by anyscalesam** [comment]
@GabeChurch do you have a new PR here yet?

**Comment by GabeChurch** [comment]
> @GabeChurch do you have a new PR here yet?

Yes, it is https://github.com/ray-project/ray/pull/45354

**Pull request closed without merging**